### PR TITLE
feat: Add utility to prepare datastore migrations

### PR DESCRIPTION
**Pull request by rhybrillou**
## Description

Writing new datastore migrations follows a few mandatory manual steps that can easily be overlooked.
The point of this change is to introduce a code generator that automates the first few steps and leaves a task list closer to where the migration code is being written

With this PR, the management around the migration itself is automated with a makefile target:
`DESCRPTION="short description of what the new migration does" make  bootstrap_migration`

## Checklist
- [ ] Investigated and inspected CI test results
~~- [ ] Unit test and regression tests added~~
~~- [ ] Evaluated and added CHANGELOG entry if required~~
~~- [ ] Determined and documented upgrade steps~~
~~- [ ] Documented user facing changes (create PR based on [openshift/openshift-docs](https://github.com/openshift/openshift-docs) and merge into [rhacs-docs](https://github.com/openshift/openshift-docs/tree/rhacs-docs))~~

If any of these don't apply, please comment below.

## Testing Performed

Manual testing:
Run make with the new Makefile target.

### CODE REVIEW

**Comment by rhybrillou** [comment]
Current dependencies on/for this PR:
* master
  * **PR #5193** <a href="https://app.graphite.dev/github/pr/stackrox/stackrox/5193" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/stackrox/stackrox/5193?utm_source=stack-comment).

File: Makefile [hunk 1]
```diff
@@ -802,3 +802,7 @@ mitre:
 	@echo "+ $@"
 	CGO_ENABLED=0 GOOS=$(HOST_OS) $(GOBUILD) ./tools/mitre
 	go install ./tools/mitre
+
+.PHONY: bootstrap_migration
+bootstrap_migration:
```
Comment (janisz) [inline Makefile:807]: Please add a README entry for that.

File: tools/generate-helpers/bootstrap-migration/migration_impl.go.tpl [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+//  - remove the calls to GetSchemaForTable and to RegisterTable from the copied schema files
+//  - remove the xxxTableName constant from the copied schema files
+//  - copy the newly generated stores from their location in central to appropriate subdirectories in stores/updated
+//  - remove any unused function from the copied store files (the minimum for the public API should contain Walk, UpsertMany, DeleteMany)
+//  - remove the scoped access control code from the copied store files
+//  - remove the metrics collection code from the copied store files
+
+func migrate(database *types.Databases) error {
+    ctx := sac.WithAllAccess(context.Background())
+    _ = ctx
```
Comment (janisz) [inline tools/generate-helpers/bootstrap-migration/migration_impl.go.tpl:30]:
```suggestion
    _ = ctx //TODO: remove it, it's just to make compiler happy
```

File: tools/generate-helpers/bootstrap-migration/main.go [hunk 3]
```diff
@@ -0,0 +1,248 @@
...
+func renderFile(templateMap map[string]interface{}, temp func(s string) *template.Template, templateFileName string) error {
+	buf := bytes.NewBuffer(nil)
+	if err := temp(templateFileName).Execute(buf, templateMap); err != nil {
+		return err
+	}
+	file := buf.Bytes()
+
+	importProcessingStart := time.Now()
+	formatted, err := imports.Process(templateFileName, file, nil)
+	importProcessingDuration := time.Since(importProcessingStart)
```
Comment (janisz) [inline tools/generate-helpers/bootstrap-migration/main.go:214]: I think it's better to keep template in good shape than calling `imports`
```suggestion
```

File: tools/generate-helpers/bootstrap-migration/main.go [hunk 4]
```diff
@@ -0,0 +1,248 @@
...
+	isInImports := false
+	registered := false
+	for fileScanner.Scan() {
+		line := fileScanner.Text()
+		if strings.HasPrefix(line, "import (") {
+			isInImports = true
+		}
+		if isInImports && strings.HasPrefix(line, ")") {
+			if !registered {
+				registrationPrefix := "github.com/stackrox/rox/migrator/migrations"
```
Comment (janisz) [inline tools/generate-helpers/bootstrap-migration/main.go:170]: move to const

File: tools/generate-helpers/bootstrap-migration/main.go [hunk 5]
```diff
@@ -0,0 +1,248 @@
...
+	defer closeFile(readFile)
+	if err != nil {
+		return nil, err
+	}
+	fileScanner := bufio.NewScanner(readFile)
+	isInImports := false
+	registered := false
+	for fileScanner.Scan() {
+		line := fileScanner.Text()
+		if strings.HasPrefix(line, "import (") {
```
Comment (janisz) [inline tools/generate-helpers/bootstrap-migration/main.go:165]: Maybe add a comment on top of `all.go` to indicate it's generated file

Comment (janisz) [reply tools/generate-helpers/bootstrap-migration/main.go:165]:
```
// Code generated by make bootstrap_migration . DO NOT EDIT.
```

File: tools/generate-helpers/bootstrap-migration/seq_num.go.tpl [hunk 6]
```diff
@@ -0,0 +1,11 @@
+package internal
```
Comment (janisz) [inline tools/generate-helpers/bootstrap-migration/seq_num.go.tpl:1]: add generate comment to indicate this file should not be changed

**Review by janisz — COMMENTED** [review]
just minor comments about adding generated file comments and readme entry

File: tools/generate-helpers/bootstrap-migration/seq_num.go.tpl [hunk 7]
```diff
@@ -0,0 +1,18 @@
...
+	// CurrentDBVersionSeqNum is the current DB version number.
+	// This must be incremented every time we write a migration.
+	// It is a shared constant between central and the migrator binary.
+	CurrentDBVersionSeqNum = {{.nextSeqNum}}
+
+	// MinimumSupportedDBVersionSeqNum is the minimum DB version number
+	// that is supported by this database.  This is used in case of rollbacks in
+	// the event that a major change introduced an incompatible schema update we
+	// can inform that a rollback below this is not supported by the database
+	MinimumSupportedDBVersionSeqNum = {{.nextSeqNum}}
```
Comment (dashrews78) [inline tools/generate-helpers/bootstrap-migration/seq_num.go.tpl:14]: Don't automatically update this one.  Should probably add it to one of the TODOs to evaluate if it needs to be updated.

File: tools/generate-helpers/bootstrap-migration/seq_num.go.tpl (hunk 6 above)
Comment (dashrews78) [reply tools/generate-helpers/bootstrap-migration/seq_num.go.tpl:2]: I'm not sure we can have this totally generated unless we provide a mechanism for passing in `MinimumSupportedDBVersionSeqNum` as that will diverge from `{{.nextSeqNum}}` and should in fact rarely be updated.

**Review by janisz — APPROVED** [review]

**Review by dashrews78 — APPROVED** [review]

**Pull request merged**